unlock-app/src/components/interface/keychain/CancelAndRefundModal.tsx

```diff
@@ -1,84 +1,100 @@
+import { Web3Service } from '@unlock-protocol/unlock-js'
+import React, { useEffect, useState } from 'react'
+import { useQuery } from 'react-query'
 import { useWalletService } from '~/utils/withWalletService'
 import { ToastHelper } from '../../helpers/toast.helper'
-import InlineModal from '../InlineModal'
-import Loading from '../Loading'
 
 export interface ICancelAndRefundProps {
   active: boolean
   lock: any
-  dismiss: () => void
+  setIsOpen: (status: boolean) => void
   account: string
   currency: string
   keyId: string
+  network: number
 }
 
+const CancelAndRefundModalPlaceHolder = () => {
+  return (
+    <div className="flex flex-col w-full gap-5 p-4">
+      <div className="flex flex-col gap-2">
+        <div className="h-[24px] w-2/3 bg-slate-200 animate-pulse"></div>
+        <div className="h-[14px] w-1/2 bg-slate-200 animate-pulse"></div>
+      </div>
+      <div className="h-[50px] w-full rounded-full bg-slate-200 animate-pulse"></div>
+    </div>
+  )
+}
+
+const MAX_TRANSFER_FEE = 10000
+
 export const CancelAndRefundModal: React.FC<ICancelAndRefundProps> = ({
   active,
   lock,
-  dismiss,
+  setIsOpen,
   account: owner,
   currency,
   keyId,
+  network,
 }) => {
-  const [loading, setLoading] = useState(false)
-  const [loadingAmount, setLoadingAmount] = useState(false)
-  const [refundAmount, setRefundAmount] = useState('')
+  const [isRefundable, setIsRefundable] = useState(false)
+  const [hasMaxCancellationFee, setHasMaxCancellationFee] = useState(false)
   const walletService = useWalletService()
   const { address: lockAddress, tokenAddress } = lock ?? {}
 
-  useEffect(() => {
+  const getRefundAmount = async () => {
     if (!active) return
-    const getRefundAmount = async () => {
-      setLoadingAmount(true)
-      const params = {
-        lockAddress,
-        owner,
-        tokenAddress,
-        tokenId: keyId,
-      }
-      const totalToRefund = await walletService.getCancelAndRefundValueFor(
-        params,
-        () => true
-      )
-      setRefundAmount(totalToRefund)
-      setLoadingAmount(false)
+
+    const params = {
+      lockAddress,
+      owner,
+      tokenAddress,
+      tokenId: keyId,
     }
-    getRefundAmount()
-  }, [
-    active,
-    setRefundAmount,
-    setLoadingAmount,
-    lockAddress,
-    tokenAddress,
-    keyId,
-    owner,
-    walletService,
-  ])
-
-  const onCloseCallback = () => {
-    if (typeof dismiss === 'function') dismiss()
-    setLoading(false)
+    if (!walletService) return
+    return await walletService.getCancelAndRefundValueFor(params, () => true)
   }
 
-  const onCancelAndRefund = async () => {
-    setLoading(true)
+  const getCancellationFee = async () => {
+    const web3Service = new Web3Service(networks)
+    return await web3Service.transferFeeBasicPoints(lockAddress, network)
+  }
 
+  const getLockDetails = async () => {
+    const web3Service = new Web3Service(networks)
+    return await web3Service.getLock(lock.address, network)
```

Comment:
I worry this might be slow because it retrieves a lot of data we won't really need here. Can you just get the balance (maybe it's worth adding a `getLockBalance` function to web3Service for this (and refactor `parseLockGetters` to use it)

Reply:
updated with `getLock` -> `getAddressBalance` is already present in `web3Service`.
Is also present inside `parseLockGetters` to get lock balance.